src/sources/models.js

```diff
@@ -3,14 +3,67 @@ const {fetch} = require("../fetch");
 const {NoResourcePathError} = require("../exceptions");
 const utils = require("../utils");
 
-
-/* These Source, Dataset, and Narrative classes contain information to map an
- * array of dataset/narrative path parts onto a URL.  Source selection and
- * dataset path aliasing (/flu → /flu/seasonal/h3n2/ha/3y) is handled in
- * utils/prefix.parsePrefix().
+/* The model classes here are the base classes for the classes defined in
+ * ./core.js, ./community.js, ./groups.js, etc.
+ *
+ * Sources represent a remote HTTP data source.  Sources contain Datasets and
+ * Narratives (both Resources).  Datasets and Narratives have Subresources
+ * representing specific files/data that make up the conceptual Resource as a
+ * whole.
+ *
+ * Source
+ *   Dataset                (implements Resource interface)
+ *     DatasetSubresource   (implements Subresource interface)
+ *   Narrative              (implements Resource interface)
+ *     NarrativeSubresource (implements Subresource interface)
+ *
+ * Subresources are separate from Resources so that our code can more easily
+ * talk about, address, and pass around values representing both the conceptual
+ * whole of a dataset or narrative and the concrete individual files making up
+ * those wholes.
+ *
+ * A concrete example:
+ *
+ * CoreSource (in ./core.js) represents a Cloudfront distribution
+ * (https://data.nextstrain.org) in front of an S3 bucket
+ * (s3://nextstrain-data).
+ *
+ *   const coreSource = new CoreSource()
+ *
+ * The core dataset "flu/seasonal/h3n2/ha/2y" is represented by a Dataset
+ * instance you get by calling:
+ *
+ *   const dataset = coreSource.dataset(["flu", "seasonal", "h3n2", "ha", "2y"])
+ *
+ * That Dataset has Subresources identified by the names "main" and
```

Comment:
I'm interested in bigger picture architecture and it's really nice to see the server evolve from a collection of functions which had to pass around lots of data to middeware which instantiates classes representing the requested data. 

With that in mind, what was the motivation to have a `Subresource` class rather than a method on (e.g.) the Resource object? I know that dataset's are complicated as one dataset can have multiple assets (files) which complicates methods such as `dataset.exists()` - was this part of the motivation?

Reply:
The reason for the Resource and Subresource split is precisely because of the distinction between the conceptual whole ("a dataset" or "a narrative") and the concrete data/files/parts that comprise the whole. The codebase needs to be able to track/query information about both the whole and the parts and also needs to be able to pass representations of them around. I wrote this paragraph to try to cover just that point:

https://github.com/nextstrain/nextstrain.org/blob/85305017e5519ad37ced55c0fdefb7f79d23ede0/src/sources/models.js#L20-L23

The example of `dataset.exists()` is precisely an example of this distinction, but also things like `dataset.resolve()` and, in #443, things like `subresource.media_type`.